### PR TITLE
feat(badge): added leaderboard for badges

### LINKED ISSUE #306: Badges ledertavle
Som medlem i TIHLDE ønsker jeg å se hvem som har flest badges totalt sett, hvem som har flest badges per årstrinn og hvem som har et spesifikt badge. Dette bør kun være tilgjengelig for innloggede medlemmer. Man trenger her endepunkter for de forskjellige listene, eller at det gjøres gjennom et BadgesFilter

### PR DESCRIPTION
**Pull request by CJGutz**
## Proposed changes

This PR will bring a badge leaderboard and an overview of which users have which badges. There are two leaderboards. The first one shows a list of users with how many badges they each have. This leaderboard will not show which specific badges the user has acquired. This is the main leaderboard and is ordered by the number of badges a user has or their first name. The second, the badge-specific leaderboard, lists users who have a specific badge. It will be ordered by what date the badge was found starting with the person that got the badge first. Frontend might add the possibility for when clicking on a user either 1. shows what badges the user has in a dropdown menu or 2.  brings them to the user's profile. The ladder is in line with public profiles #324 . 

Going to a user's profile and choosing to look at badges will only bring up public badges or badges that the visiting user, the user sending the get-request, has achieved. A badge is public if the active_to date has been passed or if active_to is set to null and the active_from date has been passed or if active_to and active_from is null. The badge-specific leaderboard doesn't have to be filtered by public. The badge-specific leaderboard should only be found if a member already has the id. Regarding the main leaderboard and the number-of-badges number, all badges can be counted. The important bit is that the specific badge isn't shown to a member who hasn't unlocked it yet.

Tests to these leaderboards have been added. Both integration and model tests. 

Issue number: closes #306 

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] CHANGELOG.md has been updated. [Guide](https://tihlde.slab.com/posts/changelog-z8hybjom)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] The fixtures have been updated if needed (for migrations)
- [x] Pull request title follows [conventional commits](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) (`type(scope): description`)
- [x] Format and lint (`make format` and `make flake8`) was run locally without failures

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

For the urls `/badges/`, `/badges/leaderboard/` and `/badges/<badge_id>/leaderboard/` I have used different models.
When showing all badges I have used the Badge model. I have set the permission_classes to isHS so that not everyone can see all badges. 
For the main leaderboard, I use the User model since all Users should be shown if they have more than one badge. Filters and search functions are also added. Filters are based on class, study, and the category of a badge. The search functionality uses the user's user_id, first name, and last name.
The badge-specific leaderboard uses the UserBadge model. Since only one user can have one specific badge, there will not be several users on the leaderboard. The queryset shows only UserBadges with a corresponding id to badge in the url.

The UserBadgeViewSet has been drastically changed and has been moved to UserViewSet as a action method. URL for this method is `users/<user_id>/badges/`. A get request will return badges of the specified user. When user_id is set to "me", the user sending the request will be chosen. A post request will only make a badge for requesting user even though user_id is set to someone else's user_id.

A new app badge has been added. It contains the models badge, user_badge and badge_category.

Integration tests check different status codes from the response for different actions. Actions include, but are not limited to, get-request for an anonymous user and post-requests for members. Model tests focus on testing if a badge is active or public.
In both cases, I use testing parameters for easier readability and writing shorter code while getting the same results.

### CODE REVIEW

File: app/content/serializers/user_badge.py [hunk 1]
```diff
@@ -31,3 +31,28 @@ def get_badge(self, obj):
...
+        return {
+            "user_id": obj.user_id,
+            "first_name": obj.first_name,
+            "last_name": obj.last_name,
+            "user_class": obj.user_class,
+            "user_study": obj.user_study,
+        }
+
+    def get_number_of_badges(self, obj):
+        return len(UserBadge.objects.filter(user=obj))
```
Comment (olros) [inline app/content/serializers/user_badge.py:58]: Bruk `.count()` etter filter istedenfor `len()` rundt. Da telles det med sql istedenfor python, noe som er myyyye raskere

File: app/content/views/leaderboard.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+from app.content.models.user_badge import UserBadge
+from app.content.serializers.badge import BadgeSerializer
+from app.content.serializers.user_badge import LeaderboardSerializer
+
+
+class LeaderboardViewSet(viewsets.ModelViewSet):
+    queryset = (
+        User.objects.all()
+        .annotate(number_of_badges=Count("user_badges"))
+        .order_by("-number_of_badges")
```
Comment (olros) [inline app/content/views/leaderboard.py:21]: Du trenger ikke annontere her i tillegg til å kjøre en egen spørring for det samme i serializeren

**Review by olros — CHANGES_REQUESTED** [review]

File: app/content/views/leaderboard.py (hunk 2 above)
Comment (CJGutz) [reply app/content/views/leaderboard.py:21]: Tanken med annotate er at man kan sortere basert på et felt som ikke kommer før i serializeren. Er det en annen bedre måte å sortere?

File: app/content/serializers/user_badge.py (hunk 1 above)
Comment (CJGutz) [reply app/content/serializers/user_badge.py:58]: ser den

File: app/content/views/leaderboard.py (hunk 2 above)
Comment (olros) [reply app/content/views/leaderboard.py:21]: Ahh, skjønner. Har ikke sånn veldig mye erfaring med slik sortering men kanskje @Eirsteir , @norgor , @Her0elt eller @trthingnes har noen tanker om hva som er lurt her.

**Comment by olros** [comment]
Prøvde å utdype alt jeg veldig gjerne har lyst på i badges-ledertavler i https://github.com/TIHLDE/Lepton/issues/306#issuecomment-973153029 nå @CJGutz . Litt av hvert å gjøre 😉

**Comment by CJGutz** [comment]
> Prøvde å utdype alt jeg veldig gjerne har lyst på i badges-ledertavler i [#306 (comment)](https://github.com/TIHLDE/Lepton/issues/306#issuecomment-973153029) nå @CJGutz . Litt av hvert å gjøre 😉

Jeg er imponert! Da har jeg litt å gjøre på fritden av eksamensperioden :).

File: app/content/serializers/user_badge.py [hunk 3]
```diff
@@ -31,3 +31,28 @@ def get_badge(self, obj):
...
+    number_of_badges = serializers.SerializerMethodField()
+
+    class Meta:
+        model = User
+        fields = (
+            "user",
+            "number_of_badges",
+        )
+
+    def get_user(self, obj):
```
[the remark this answers is not in the document]
Comment (CJGutz) [reply app/content/serializers/user_badge.py:47]: Hvordan er det vi da får tilgang på studie og klasser? Jeg så at strikes også bruker DefaultUserSerializer, som ikke har klasse og studie, men får opp studie og klasse likevel på nettsiden. Er det noe dere gjør i frontend?

[the remark this answers is not in the document]
Comment (olros) [reply app/content/serializers/user_badge.py:47]: Hvis jeg husker riktig så henter vi ut oversikten over brukere med prikker gjennom `/user/?has_active_strikes=true` og da får man hele brukeren.

Jeg tenker det går helt fint om du legger til user_class og user_study i DefaultUserSerializer, er ikke noe sensitiv personopplysning tenker jeg

[the remark this answers is not in the document]
Comment (CJGutz) [reply app/content/serializers/user_badge.py:47]: Dette er hvordan jeg har løst å få med DefaultSerializer. Jeg har prøvd mange ulike med det virker som om dette var den eneste metoden som funket. Jeg så at submissions-serializer hadde fått til nested serializer med DefaultUserSerializer, som er det jeg ønsker, men jeg får det ikke til med samme metode. Jeg fikk da bare all informasjon på samme nivå.
Men denne metoden er ikke så verst heller,
```
class UserBadgeLeaderboardSerializer(BaseModelSerializer):
    user = serializers.SerializerMethodField()
    
    class Meta:
        model = UserBadge
        fields = ("user", "created_at")

    def get_user(self, obj):
        return DefaultUserSerializer(obj.user).data
```

File: app/content/views/user.py [hunk 4]
```diff
@@ -137,12 +139,55 @@ def get_user_memberships(self, request, *args, **kwargs):
...
+        user = self.request.user
+        badge = get_object_or_404(Badge, flag=request.data.get("flag"))
+
+        if not badge.active:
+            return Response(
+                {"detail": "Badgen er ikke aktiv"}, status=status.HTTP_400_BAD_REQUEST,
+            )
+        serializer = UserBadgeSerializer(data=request.data)
+        if serializer.is_valid():
+            UserBadge(user=user, badge=badge).save()
```
Comment (olros) [inline app/content/views/user.py:153]: Skal forklare på møte hvordan logging nå skal gjøres, tror det følgende er mulig men hvis ikke så kan du logge manuelt
```suggestion
            super().perform_create(serializer,user=user, badge=badge)
            # Evt:
            self._log_on_create(serializer)
```

File: app/content/views/user.py [hunk 5]
```diff
@@ -137,12 +139,55 @@ def get_user_memberships(self, request, *args, **kwargs):
...
+            user = self.get_object()
+
+        user_badges = user.user_badges.order_by("-created_at")
+        badges = [
+            user_badge.badge
+            for user_badge in user_badges
+            if user_badge.badge.is_public
+            or UserBadge.objects.filter(
+                user=request.user, badge=user_badge.badge
+            ).exists()
```
Comment (olros) [inline app/content/views/user.py:174]:
```suggestion
            or request.user.user_badges.filter(badge=user_badge.badge).exists()
```

File: app/util/exceptions.py [hunk 6]
```diff
@@ -12,7 +11,7 @@
 def exception_handler(exc, context):
     response = drf_exception_handler(exc, context)
 
-    if not response and isinstance(exc, IntegrityError) and not settings.DEBUG:
```
Comment (olros) [inline app/util/exceptions.py:15]: Fjern denne endringen, det er veldig nyttig å få hele feilmeldingen når man ikke er i prod

File: app/util/exceptions.py [hunk 7]
```diff
@@ -23,7 +22,7 @@ def exception_handler(exc, context):
     else:
         logger.error(f"Unhandled request exception: {exc}")
 
-    if not settings.DEBUG and not response:
```
Comment (olros) [inline app/util/exceptions.py:26]: Fjern denne endringen, det er veldig nyttig å få hele feilmeldingen når man ikke er i prod

File: app/tests/content/test_badge_integration.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    client = get_api_client(user=member)
+    data = {"flag": "flag_unlike_anything_else"}
+    response = client.post(url, data)
+
+    assert response.status_code == status.HTTP_404_NOT_FOUND
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(
+    ("active_from", "active_to", "expected_status_code", "expected_detail"),
```
Comment (olros) [inline app/tests/content/test_badge_integration.py:101]: Du trenger ikke sjekke `expected_detail` når `expected_status_code` i praksis sjekker det samme. I tillegg er det ikke usannsynlig at `detail` vil endres en gang, men de breaker ingenting og da trenger vi heller ikke teste det
```suggestion
    ("active_from", "active_to", "expected_status_code"),
```

File: app/tests/content/test_badge_integration.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+def _get_badge_leaderboard_url():
+    return f"{API_BADGE_BASE_URL}leaderboard/"
+
+
+def _get_badge_specific_leaderboard_url(badge):
+    return f"{API_BADGE_BASE_URL}{badge.id}/leaderboard/"
+
+
+def _get_user_badge_url():
```
Comment (olros) [inline app/tests/content/test_badge_integration.py:22]: Du henter ut badges i flertall ;)
```suggestion
def _get_user_badges_url():
```

File: app/tests/content/test_badge_integration.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+
+@pytest.mark.django_db
+def test_get_request_for_user_badges_as_anonymous_user(default_client):
+    """An anonymous user should not be able to do a get request for a user badge"""
+
+    url = _get_user_badge_url()
+    response = default_client.get(url)
+
+    assert response.status_code == status.HTTP_403_FORBIDDEN
```
Comment (olros) [inline app/tests/content/test_badge_integration.py:57]: Kan være lurt å ha samme tester for medlemmer med da forventet status_kode == 200

**Review by olros — CHANGES_REQUESTED** [review]
Nærmer seg veldig nå! Flotte med tester

**Comment by CJGutz** [comment]
> Nærmer seg veldig nå! Flotte med tester

Så bra. Gleder meg til å se det bruk. Tenker i tillegg å refactore i henhold til #332 med UserBadge, Badge og BadgeCategory inn i appen Badge

File: app/badge/migrations/0001_create_models.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+                ('title', models.CharField(max_length=200)),
+                ('description', models.CharField(max_length=200)),
+                ('active_from', models.DateTimeField(blank=True, null=True)),
+                ('active_to', models.DateTimeField(blank=True, null=True)),
+                ('badge_category', models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.SET_NULL, to='badge.badgecategory')),
+            ],
+            options={
+                'verbose_name_plural': 'Badges',
+            },
+        ),
```
Comment (olros) [inline app/badge/migrations/0001_create_models.py:51]: Å flytte en modell til en annen app er litt mer tricky enn å bare slette og delete modellen i hver app. Per nå så vil dette slette alle eksisterende badges og user-badges i databasen hvis dette hadde blitt merget. Vi kan gå gjennom dette på søndag

**Review by olros — CHANGES_REQUESTED** [review]

File: app/badge/migrations/0001_create_models.py (hunk 11 above)
Comment (CJGutz) [reply app/badge/migrations/0001_create_models.py:51]: Ja gjerne det

File: app/badge/migrations/0002_auto_20220217_1901.py [hunk 12]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 3.2.10 on 2022-02-06 13:34
+# Generated by Django 3.2.10 on 2022-02-17 18:01
```
Comment (olros) [inline app/badge/migrations/0002_auto_20220217_1901.py:1]: Gi denne filen et bedre navn ;)

File: app/content/views/user.py (hunk 5 above)
Comment (olros) [reply app/content/views/user.py:180]: Se https://github.com/TIHLDE/Lepton/pull/345/files#r796910233

File: app/util/exceptions.py [hunk 13]
```diff
@@ -1,11 +1,12 @@
 import logging
 
-from django.conf import settings
 from django.db.utils import IntegrityError
 from rest_framework import status
 from rest_framework.response import Response
 from rest_framework.views import exception_handler as drf_exception_handler
 
+from app import settings
```
Comment (olros) [inline app/util/exceptions.py:8]:
```suggestion
from django.conf import settings
from django.db.utils import IntegrityError
from rest_framework import status
from rest_framework.response import Response
from rest_framework.views import exception_handler as drf_exception_handler
```

https://stackoverflow.com/a/50839806

File: app/badge/serializers/leaderboard.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        fields = (
+            "user",
+            "number_of_badges",
+        )
+
+    def get_user(self, obj):
+        return DefaultUserSerializer(obj).data
+
+    def get_number_of_badges(self, obj):
+        return obj.user_badges.count()
```
Comment (olros) [inline app/badge/serializers/leaderboard.py:24]: Når denne brukes for ledertavle i badge-kategori så tar den ikke hensyn (såvidt jeg klarer å se) til at en kun ønsker antall badges hver bruker har innenfor kategorien. Per nå ser det ut som vi kun viser brukere som faktisk har en badge i kategorien (riktig), men så viser antall badges hver bruker har totalt blant absolutt alle badges.

Tar jeg feil?

**Review by olros — CHANGES_REQUESTED** [review]

**Review by olros — APPROVED** [review]
Synes dette ser helt topp ut! Pushet et par små endringer nå for å fikse antall badges som vises per bruker i ledertavle for kategorier. https://github.com/TIHLDE/Kvark/pull/383 skal være klar også så du/andre må gjerne teste den opp mot denne branchen.

Meget bra jobba! 🚀

**Pull request merged**